Add translate scalar function, unit tests, and docs

dvorobiov opened the pull request:
This PR adds TRANSLATE scalar string function to Crate that works like in PostgreSQL: (https://www.postgresqltutorial.com/postgresql-translate/).

The PR contains: 
- implementation of function
- unit tests
- documentation

crate-jenkins commented:
Can one of the admins verify this patch?

seut commented:
@dvorobiov Thank you a lot for this contribution, we'll have a look asap.

seut commented on sql/src/main/java/io/crate/expression/scalar/string/TranslateFunction.java:
```diff
@@ -0,0 +1,95 @@
+    private static String applyTranslationsToText(String text, HashMap<Character, Consumer<StringBuilder>> translationMap) {
+        var outputSb = new StringBuilder();
+
+        for (var ch: text.toCharArray()) {
```
This will create a copy of the `text` string. Using a for loop in conjunction with `text.charAt(i)` would avoid it.

dvorobiov replied:
right, thanks for the comment!

seut commented on sql/src/main/java/io/crate/expression/scalar/string/TranslateFunction.java:
```diff
@@ -0,0 +1,95 @@
+        var outputSb = new StringBuilder();
+
+        for (var ch: text.toCharArray()) {
+            if (translationMap.containsKey(ch)) {
```
As `containsKey` will use `get()` internally we could just call `get` and decide on the return value. This would avoid another `get` call.

seut commented on sql/src/main/java/io/crate/expression/scalar/string/TranslateFunction.java:
```diff
@@ -0,0 +1,95 @@
+    private static void checkDuplicatesInFromArg(Character ch, HashMap<Character, Consumer<StringBuilder>> translationMap) {
+        if (translationMap.containsKey(ch)) {
+            throw new IllegalArgumentException(
+                String.format(Locale.ENGLISH, "Argument 'from' contains duplicate characters '%s'", ch)
```
I wonder were this behaviour is coming from, afaik PostgreSQL does not complain on duplicate `from` chars but just ignore the later ones. Anything I'm missing?

dvorobiov replied:
Hi Sebastian!

A scenario with field values 
```
from -> to:
'aaa'-> '123'
```
can be handled by 

1. autoresolving:
1a. use the first occurence: a -> 1
1b. use the last occurence: a -> 3 (PostgreSQL)
2. raising exception

Please, let me know if it's important is to keep the implementation as close as pg's it can be a good reason to update the implementation to the scenario 1b. 
On the other hand, if it's not the priority, and the priority is a predictable, strict behaviour, the auto-resolving seems to look like an arbitrary implicit solution (in my opinion, sure).

Please, let me know your thoughts!

seut replied:
@dvorobiov Yes we want this scalar to behave the same as in PostgreSQL, so 1a please :-)

dvorobiov commented:
@seut thank you! I have just pushed amend commit with a small change for one unit test.

seut commented on sql/src/main/java/io/crate/expression/scalar/string/TranslateFunction.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    private static HashMap<Character, Consumer<StringBuilder>> createTranslationMap(String from, String to) {
+        var translationMap = new HashMap<Character, Consumer<StringBuilder>>();
```
Any reason not to use a primitive char array for the mapping? I think this would be much cheaper and faster.

seut commented:
@dvorobiov It would be nice if you could push changes as dedicated commits, this makes it easier for us to see what have changed since the last review ;-) Thanks a lot!

seut commented:
ok to test

seut commented:
@dvorobiov By playing around with this PR we recognized that the current non-optimized version (no `compile`) would even be faster.
```
cr> select max(x) from (select translate(mountain, 'on', 'wu') as x from sys.summits, generate_series(1, 10000)) t;
+--------+
| max(x) |
+--------+
| Špik   |
+--------+
SELECT 1 row in set (6.321 sec)
```
Without `compile`:
```
cr> select max(x) from (select translate(mountain, 'on', 'wu') as x from sys.summits, generate_series(1, 10000)) t;                        
+--------+
| max(x) |
+--------+
| Špik   |
+--------+
SELECT 1 row in set (5.639 sec)
```

So we think we should investigate a bit more here what solution is best e.g. using proper benchmarks.
Would you like to do that?
Otherwise we would take this PR over now if you don't mind.
In any way, thank you a lot for this contribution.

dvorobiov commented:
@seut 
Thanks for sharing your benchmarks. I wanted to do those as well, probably, on the weekend, since this week I started on a new contract and have less time. 

First, if you played with the code that I pushed yesterday in the evening, it was in progress and could not produce correct test results. I have just pushed an update where I added a few files with different translate() function implementation for testing: postgres simple implementations, optimised and unoptimised with hashmaps (I switched from Consume to Optional char in value), and my attempt to use ordered char array for lookup - I tried to understand and implement your comment.

Second, the results of your benchmark tests look suprising and interesting. Anyway, I see there are only two char long values in from-to fields. I think all those optimisations and different lookups will show more definite results for longer translates, say from 'AaBcCc...YyZz' to 'ZzYyXx...BbAa'. 

I am interested to benchmark them as well, but, as I said, I will have time on weekend only. I would like to also ask you, if you have some test dataset of a good size for that? Otherwise I ll generate something. Also, I'd like to ask you a couple qq on debugging and benchmarking :) 

if you want to take over the current code, sure, I am OK with that, if I correctly understood you question.

seut commented:
@dvorobiov 
> Thanks for sharing your benchmarks. I wanted to do those as well, probably, on the weekend, since this week I started on a new contract and have less time.

No worries, but these were not really benchmarks, just issued the queries once ;)

> Second, the results of your benchmark tests look suprising and interesting. Anyway, I see there are only two char long values in from-to fields. I think all those optimisations and different lookups will show more definite results for longer translates, say from 'AaBbCc...YyZz' to 'ZzYyXx...BbAa'.

Yes right, also I think that the most common use-case is to use a low amount of replace chars. So I would not optimize for an unusual use-case but rather for the most common one.

> I am interested to benchmark them as well, but, as I said, I will have time on weekend only.

Ok cool, there is no time pressure.

> I would like to also ask you, if you have some test dataset of a good size for that? Otherwise I ll generate something.

We have a dedicated benchmark project https://github.com/crate/crate-benchmarks which you could use. For generating fake datasets we mostly use https://github.com/mfussenegger/cr8' `insert-fake-data` command.

> Also, I'd like to ask you a couple qq on debugging and benchmarking :)

Sure anytime :) 

> if you want to take over the current code, sure, I am OK with that, if I correctly understood you question.

We really appreciate your contribution and just wanted to offer you politely to continue from here to not bother you with nitpicking related to benchmarking/performance issues going beyond this functionallity.

mfussenegger commented:
> > I would like to also ask you, if you have some test dataset of a good size for that? Otherwise I ll generate something.
> 
> We have a dedicated benchmark project https://github.com/crate/crate-benchmarks which you could use. For generating fake datasets we mostly use https://github.com/mfussenegger/cr8' `insert-fake-data` command.

I think in this case using JMH for the benchmarks might be better suited. A full end-to-end test has a lot of noise associated.


> Second, the results of your benchmark tests look suprising and interesting. Anyway, I see there are only two char long values in from-to fields. I think all those optimisations and different lookups will show more definite results for longer translates, say from 'AaBbCc...YyZz' to 'ZzYyXx...BbAa'.

Yes, the benchmark and performance characteristic will be heavily dependent on the data. It is also a question of which case we want to optimize for. There won't be a solution that outperforms other solutions in all cases. My gut feeling would be that `from` and `to` values are in many uses cases rather small (< ~100 characters).

https://en.wikipedia.org/wiki/String-searching_algorithm may be of interest here.

dvorobiov commented:
@mfussenegger thanks, I ll look into using JMH tests. 

> Yes, the benchmark and performance characteristic will be heavily dependent on the data. It is also a question of which case we want to optimize for. There won't be a solution that outperforms other solutions in all cases.

Right. I think a few parameters come into play there: 
* literals or references for "from" "to" (applicability of compile stage)
* size of "from"
* size of "text"
* size of result llist

I wanted to get a feeling how much effort to opimisiation is justifiable at Crate, and for this case.

> https://en.wikipedia.org/wiki/String-searching_algorithm may be of interest  here.

Yeah, exactly :) I am taking my copy of Cormen from the shelf :)

marregui commented:
Implemented in https://github.com/crate/crate/pull/10630
